Implement_messages_function(model,routing,controller)

tsutaya100507 opened the pull request:
# WHAT
メッセージ送信機能実装のためのモデルの作成、ルーティングの設定、アクションの定義を行い、メッセージ送信・保存ができるようにする。
# WHY
メッセージ送信機能の実装に必要なため。

Commit:
create pull request

Commit:
Create message_model and messages_table

messege_modelを作成した。また、READMEのDB設計に基づきmessages_tableを作成した。

Commit:
edit routing

メッセージ送信機能を実装するため、ルーティングを新たに記述。

Commit:
edit model for association

新たに作成したmessageモデルについて、userモデル及びgroupモデルとアソシエーションを組むための記述を追加した。

Commit:
edit controller

メッセージを新たに保存するため、messagesコントローラーにcreateアクションを定義した。

Commit:
remove new_action

メッセージ入力はindexページから直接行うこととしため、routes.rbのresourcesからnewを除外。

Commit:
install carrierwave and mini_magik

画像送信機能を実装するため、carrierwave及びmini_magickのGemをインストールした。

Commit:
set uploader

画像送信機能を実装するため、uploaderの設定を行った。

Commit:
edit_views

メッセージ送信機能を実装するため、form_forを使用した記述とし、ルーティングに合わせてビューの調整を行った。

tsutaya100507 commented:
https://gyazo.com/d02f960070ccab920a0c56d7bd39be6f

tsutaya100507 commented:
<img width="1440" alt="2017-10-11 18 33 05" src="https://user-images.githubusercontent.com/32321174/31432879-b285aaf0-aeb2-11e7-90ad-ef50c8f25218.png">

tsutaya100507 commented:
バリデーション、保存失敗時のフラッシュメッセージ、サイドバーのメッセージ表示を含めたビューの修正は次プルリクエストで実施します。

oyaoyaoyana commented on app/views/messages/index.html.haml:
```diff
@@ -0,0 +1,47 @@
+      .contents__left-content__group-list
+
+        %ul.contents__left-content__group-list__group
+          - current_user.groups.each do |group|
```
collectionオプションを使うことでeachを省くことが可能になるので調べてみてください。

oyaoyaoyana commented on app/views/messages/index.html.haml:
```diff
@@ -0,0 +1,47 @@
+                =message.message
+
+      .contents__right-content__post-space
+        = form_for @message, url: group_messages_path do |f|
```
こちら、しっかりと動いていますでしょうか？

tsutaya100507 replied:
メッセージ及び画像は送信できており、ページ遷移もできておりますが問題ありでしょうか？

yamamotokeishiro replied:
url: group_messages_path
ここの部分は削除すると動かなくなるでしょうか？

tsutaya100507 replied:
削除すると、存在しないmessages_pathに送る形になりエラーが出てしまいます。createアクションを呼ぶためのパス（ヘルパー？）としてgroup_messages_pathを指定して解決したのですが、やはり妙ですか？？
<img width="1440" alt="2017-10-12 12 33 34" src="https://user-images.githubusercontent.com/32321174/31478415-270c3ca2-af4b-11e7-9606-0ee3c9d04ca5.png">

yamamotokeishiro replied:
form_forは自動でcreateかupdateかを判断してくれます！
なのでform_forにパス指定する場合は、create,update以外のアクションに飛ばしたい場合のみに使用することが自然ですね！

oyaoyaoyana reviewed (commented):
コメント致しました。

Commit:
correct views for using collection_option of partial

eachでグループ名を表示していたところを部分テンプレートのcollectionオプションを使用して繰り返し表示できるように修正した。

Commit:
Correct how to write "form_for"

ネストしている場合のform_forの書き方に修正しました。

takuya032 commented on app/views/messages/index.html.haml:
```diff
@@ -42,7 +40,7 @@
 
       .contents__right-content__post-space
-        = form_for @message, url: group_messages_path do |f|
+        = form_for [@group,@message] do |f|
```
good!

tsutaya100507 replied:
ありがとうございます！

takuya032 commented on app/views/groups/_groups.html.haml:
```diff
@@ -1,3 +1,4 @@
-  %p まだメッセージはありません。
+=link_to(group_messages_path(groups.id)) do
+  %li.contents__left-content__group-list__group__list
+    %h3= groups.name
```
名前が適切ではありませんね。
ここで表示されているのは、groupですよね？

takuya032 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,6 +1,24 @@
   def index
+    @group = Group.find(params[:group_id])
+    @message = Message.new
+    @messages =@group.messages.order('id DESC')
```
ここ= の後ろスペース入れてください

tsutaya100507 replied:
修正しました。

takuya032 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,6 +1,24 @@
+  def create
+    @message = Message.new(message_params)
+    if @message.save
+      redirect_to group_messages_path
```
このパス引数なしで使うのおかしいですね
これどこに飛んでますか？

takuya032 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,6 +1,24 @@
 class MessagesController < ApplicationController
 
   def index
+    @group = Group.find(params[:group_id])
```
これcreateでも良いしてあげると使いやすいと思うのbefore_actionで切り離してあげてください

tsutaya100507 replied:
修正しました。

takuya032 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,6 +1,24 @@
+  end
+
+  def create
+    @message = Message.new(message_params)
```
アソシエーションを使って書いちゃいましょう。
したの、ストロングパラメーターの項目一つへらせそうですね

tsutaya100507 replied:
修正しました！

ghost commented:
デプロイされているchat-spaceと同じようにアイコンクリックで写真で送れるようにしてください。
そして、もう一度gifお願いします

tsutaya100507 commented:
https://gyazo.com/485725240d35b0efd4fe747e6d941f4c

tsutaya100507 commented:
画像の表示、バリデーション、保存失敗時のフラッシュメッセージ、サイドバーのメッセージは次プルリクエストで実施します。

Commit:
Correct messages_controller

createアクションのパスの指定の方法（引数を追記）を修正。アソシエーションを使用した@messageを定義。ストロングパラメーターからの不要
となった記述を削除。重複記述をbefore_actionとしてまとめた。

Commit:
Correct_view

画像アイコンをクリックして投稿画像の選択ができるように修正した。
